File: NationalPokedex/NationalPokedex/Models/ApiError.swift
//
//  ApiError.swift
//  NationalPokedex
//
//  Created by Akihiro Matsuyama on 2024/04/05.
//

import Foundation

/// APIを叩く際に発生する可能性のあるエラー一覧
enum ApiError: Error {
    case urlError
    case responseError(Error)
    case responseDataEmpty
    case jsonParseError(String)

    var localizedDescription: String {
        switch self {
        case .urlError: return "URL変換失敗"
        case .responseError(let error): return "APIレスポンスエラー　詳細: （\(error)）"
        case .responseDataEmpty: return "APIのレスポンスデータがnil"
        case .jsonParseError(let message): return "JSONのパースに失敗しました。失敗したデータ: (\(message)"
        }
    }
}

File: NationalPokedex/shared/src/commonMain/kotlin/com/akidon0000/nationalpokedex/Api/ApiError.kt
package com.akidon0000.nationalpokedex.Api

import kotlinx.serialization.Serializable

@Serializable
sealed class ApiError : Throwable() {
    object UrlError : ApiError()
    data class ResponseError(val error: Throwable) : ApiError()
    object ResponseDataEmpty : ApiError()
    data class JsonParseError(val json: String) : ApiError()

    override val message: String?
        get() = when (this) {
            is UrlError -> "URL変換失敗"
            is ResponseError -> "APIレスポンスエラー 詳細: (${error.toString()})"
            is ResponseDataEmpty -> "APIのレスポンスデータがnil"
            is JsonParseError -> "JSONのパースに失敗しました。失敗したデータ: ($json)"
        }
}
